src/metabase/sync/sync_metadata/fields.clj

```diff
@@ -137,29 +183,45 @@
+          (when (= (canonical-name field-metadata)
+                   (canonical-name other-field-metadata))
               other-field-metadata))
         other-metadata))
 
+(declare sync-field-instances!)
+
+(s/defn ^:private update-field-chunk!
+  [table :- i/TableInstance, known-fields :- {s/Str TableMetadataFieldWithID}, fields-to-update :- [i/TableMetadataField]]
+  (doseq [our-field fields-to-update
+          :let [db-field (get known-fields (canonical-name our-field))]]
+    ;; if field exists in both metadata sets then make sure the data recorded about the Field such as database_type is
+    ;; up-to-date...
+    (update-field-metadata-if-needed! table db-field our-field)
+    ;; ...and then recursively check the nested fields.
+    (when-let [db-nested-fields (seq (:nested-fields our-field))]
+      (sync-field-instances! table (set db-nested-fields) (:nested-fields db-field) (:id our-field)))))
+
 (s/defn ^:private sync-field-instances!
   "Make sure the instances of Metabase Field are in-sync with the DB-METADATA."
   [table :- i/TableInstance, db-metadata :- #{i/TableMetadataField}, our-metadata :- #{TableMetadataFieldWithID}
    parent-id :- ParentID]
-  ;; Loop thru fields in DB-METADATA. Create/reactivate any fields that don't exist in OUR-METADATA.
-  (doseq [db-field db-metadata]
-    (sync-util/with-error-handling (format "Error checking if Field '%s' needs to be created or reactivated"
-                                           (:name db-field))
-      (if-let [our-field (matching-field-metadata db-field our-metadata)]
-        ;; if field exists in both metadata sets then...
-        (do
-          ;; ...make sure the data recorded about the Field such as database_type is up-to-date...
-          (update-field-metadata-if-needed! table our-field db-field)
-          ;; ...and then recursively check the nested fields.
-          (when-let [db-nested-fields (seq (:nested-fields db-field))]
-            (sync-field-instances! table (set db-nested-fields) (:nested-fields our-field) (:id our-field))))
-        ;; otherwise if field doesn't exist, create or reactivate it
-        (create-or-reactivate-field! table db-field parent-id))))
+  (let [known-fields (u/key-by canonical-name our-metadata)]
+    ;; Loop thru fields in DB-METADATA. Create/reactivate any fields that don't exist in OUR-METADATA.
+    (doseq [db-field-chunk (partition-all 1000 db-metadata)]
+      (sync-util/with-error-handling (format "Error checking if Fields '%s' needs to be created or reactivated"
+                                             (map :name db-field-chunk))
+        (let [known-field-pred (comp known-fields canonical-name)
+              fields-to-update (filter known-field-pred db-field-chunk)
+              new-fields       (remove known-field-pred db-field-chunk)]
+
+          (update! existing-fields-hist (count fields-to-update))
+          (update! new-fields-hist      (count new-fields))
```

Comment:
On average, how many fields are we updating vs. finding as new

![new-fields](https://user-images.githubusercontent.com/215910/34177324-dc98db68-e4c8-11e7-805f-71a99dd91923.png)

![existing-fields](https://user-images.githubusercontent.com/215910/34177328-e0e514d4-e4c8-11e7-9014-05bb12a4d564.png)